qiskit_experiments/framework/experiment_data.py

```diff
@@ -767,6 +770,49 @@ def add_data(
     ) -> None:
         """Add experiment data.
 
+        Args:
+            data: Experiment data to add. Several types are accepted for convenience:
+
+                * Result: Add data from this ``Result`` object.
+                * List[Result]: Add data from the ``Result`` objects.
+                * Dict: Add this data.
+                * List[Dict]: Add this list of data.
+
+        Raises:
+            TypeError: If the input data type is invalid.
+        """
+
+        if any(not future.done() for future in self._analysis_futures.values()):
+            LOG.warning(
+                "Not all analysis has finished running. Adding new data may "
+                "create unexpected analysis results."
+            )
+        if not isinstance(data, list):
+            data = [data]
+
+        # Directly add non-job data
+        with self._result_data.lock:
+            for datum in data:
+                if isinstance(datum, dict):
+                    if datum["metadata"].get("composite_metadata"):
+                        tmp_exp_data = ExperimentData()
+                        marginalized_data = self._marginalized_component_data([datum])
+                        for inner_datum in marginalized_data:
+                            tmp_exp_data.__add_data(inner_datum)
+                        self._set_child_data([tmp_exp_data])
+                    else:
+                        self._result_data.append(datum)
+                elif isinstance(datum, Result):
+                    self.__add_data(datum)
+                else:
+                    raise TypeError(f"Invalid data type {type(datum)}.")
+
+    def __add_data(
```

Comment:
I don't think subroutine is necessary. You can just recursively call `add_data` as long as you find `composite_metadata` in the circuit metadata dictionary.